client/client/main.go

```diff
@@ -0,0 +1,151 @@
+// `lastIndexed`.
+func addAllFiles(cli *client.Client, lastIndexed time.Time) filepath.WalkFunc {
+	return func(path string, info os.FileInfo, err error) error {
+		if info.IsDir() && (info.Name()[0] == '.' || info.ModTime().Before(lastIndexed)) {
+			return filepath.SkipDir
+		} else if !info.IsDir() && info.Name()[0] != '.' {
+			if info.ModTime().After(lastIndexed) {
+				cli.AddFile(path)
+				fmt.Println("Added:", path)
+			}
+		}
+		return nil
+	}
+}
+
+// periodicAdd scans the files in the client directory every minute and adds
+// the updated files to the search server.
+func periodicAdd(cli *client.Client) {
+	for {
+		currTime := time.Now()
+
+		var lastIndexed time.Time
+		lastIndexedJSON, err := ioutil.ReadFile(filepath.Join(*clientDirectory, ".timestamp"))
+		if err == nil {
+			if parseErr := lastIndexed.UnmarshalJSON(lastIndexedJSON); parseErr != nil {
+				panic(fmt.Sprintf("Error when accessing the last indexed timestamp: %s", err))
+			}
+		} else if !os.IsNotExist(err) {
+			panic(fmt.Sprintf("Error when accessing the last indexed timestamp: %s", err))
+		}
+
+		if err := filepath.Walk(*clientDirectory, addAllFiles(cli, lastIndexed)); err != nil {
+			panic(fmt.Sprintf("Error when indexing the files: %s", err))
+		}
+
+		currTimeJSON, err := currTime.MarshalJSON()
+		if err != nil {
+			panic(fmt.Sprintf("Error when writing the timestamp: %s", err))
+		}
+
+		ioutil.WriteFile(filepath.Join(*clientDirectory, ".timestamp"), currTimeJSON, 0666)
+
+		fmt.Printf("\n[%s]: All files Indexed\n", currTime.Format("2006-01-02 15:04:05"))
+		time.Sleep(time.Second * 60)
+	}
+}
+
+// fetchMasterSecret fetches the master secret from the client directory.  If
+// the master secret is not present, a new one is generated and written to the
+// directory.  An error is returned if the there is an issue accessing the
+// master secret or the master secret is of the wrong length.
+func fetchMasterSecret() ([]byte, error) {
+	masterSecret, err := ioutil.ReadFile(filepath.Join(*clientDirectory, ".secret"))
+	if os.IsNotExist(err) {
+		masterSecret = make([]byte, *lenMS)
+		if _, err := rand.Read(masterSecret); err != nil {
+			return nil, err
+		}
+		if err := ioutil.WriteFile(filepath.Join(*clientDirectory, ".secret"), masterSecret, 0666); err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	} else if len(masterSecret) != *lenMS {
+		return nil, errors.New("Invalid master secret length")
+	}
+	return masterSecret, nil
+}
+
+// performSearchWord searched fot the word `keyword` on `cli`, and print out the
+// results.
+func performSearchWord(cli *client.Client, keyword string) {
+	files, err := cli.SearchWord(keyword)
+	if err != nil {
+		fmt.Printf("Error when searching word %s: %s", keyword, err)
+		return
+	}
+	args := make([]string, len(files)+2)
+	args[0] = "-ilZw"
+	args[1] = keyword
+	copy(args[2:], files[:])
+	output, _ := exec.Command("grep", args...).Output()
```

Comment:
The grep feels like it should do in `SearchWord`, because otherwise we'll have to re-implement it for every type of client right?  (For example, a GUI that uses `client.Client` directly.)